vllm/v1/worker/gpu_model_runner.py

```diff
@@ -799,6 +799,21 @@ def execute_model(
                     # This relies on cuda-specific torch-internal impl details
                     generator.set_offset(generator.get_offset() - 4)
 
+        # num_reqs entries should be non-None
+        assert all(
+            req_id is not None for req_id in
+            self.input_batch.req_ids[:num_reqs]), "req_ids contains None"
+        req_ids = cast(List[str], self.input_batch.req_ids[:num_reqs])
+
+        # NOTE: GPU -> CPU Sync happens here.
+        # Move as many CPU operations as possible before this sync point.
+        sampled_token_ids = sampler_output.sampled_token_ids.tolist()
```

Comment:
It might be faster to skip `.tolist()` and instead do `sampler_output.sampled_token_ids.cpu()` and then `sampler_output.sampled_token_ids[i].item()` in the inner loop.

Reply:
In my experience, `item()` took considerable time so should be avoided.